keps/sig-storage/20191016-csi-conformance.md

```diff
@@ -0,0 +1,148 @@
+---
+title: kubernetes-extensions-suite-for-verifying-optional-components  (CSI Verification Test Suite)
+authors:
+  - "@brahmaroutu"
+  - "@bradtopol"
+owning-sig: sig-architecture, sig-storage
+participating-sigs:
+  - sig-storage
+  - sig-architecture
+reviewers:
+  - "@timothysc"
+  - "@johnbelamaric"
+approvers:
+  - "@bgrant0607"
+  - "@msau42"
+editor: "@brahmaroutu"
+creation-date: 2019-10-16
+last-updated: 2019-10-16
+status: implementable
+---
+
+# kubernetes-extensions-suite-for-verifying-optional-components (CSI Verification Test Suite)
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Phase 1](#phase-1)
+  - [Phase 2](#phase-2)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Annotate test files with behaviors](#annotate-test-files-with-behaviors)
+  - [Annotate existing API documentation with behaviors](#annotate-existing-api-documentation-with-behaviors)
+  - [Manual / Periodic Testing](#manual--periodic-testing)
+- [Pros](#pros)
+- [Cons](#cons)
+<!-- /toc -->
+
+## Summary
+
+Need to define a suite of tests to properly provide coverage for CSI specification.
+
+Proposal to enable a set of guidelines that SHALL be followed when building e2e tests for optional components such that they can be later promoted into a verification test suite. Vendors will use these test suites to test compliance to the additional behavior defined by the component specification. At this moment the test suite will help vendors ability to test their implementation, whether is meets the specification.
```

Comment:
Sorry for the delay in response, this fell off my radar.

I actually think there are multiple orthogonal goals:
1) Validate that Kubernetes distributions support the CSI spec. This should be a requirement for all Kubernetes distributions.
2) Validate that storage vendors' CSI drivers are integrated well with Kubernetes. This is helpful for vendors developing CSI drivers.
3) Validate that a Kubernetes distribution provides a good default out of the box storage integration. This could be a candidate for an optional profile.

I think that for the Kubernetes project, 1) is most important, while 2) we already have, and 3) is a nice to have and maybe we can consider after the first 2.

Reply:
Thanks for the response. I would not tackle option 3, but would like some direction around option 1.
If I understand this correctly, sanity test suite(option 2) is a way to test CSI Spec. Unless we have variations in a mock driver to test CSI API with different capabilities we do not have a list of drivers to cover all API from CSI Spec.  Should we somehow run a mock driver across various Kubernetes distributions and run e2e? Should we start developing mock behaviors?

Reply:
We have some mock driver tests already under csi_mock_volumes.go which I think would be a good candidate for 1). I started working on promoting one test, but found a bug, so it is stalled at the moment until it is fixed. https://github.com/kubernetes/kubernetes/pull/85164.  There are additional tests in that file that I think could be good candidates to promote to conformance: https://github.com/kubernetes/kubernetes/issues/85101#issuecomment-608153902

Reply:
Thanks for the review and pointers to the issue. I can work on some of the tests.